### PR TITLE
clojure 1.10.1.561

### PR DESCRIPTION
**Pull request by puredanger**
The Clojure team releases new versions of the clj tools frequently (sometimes several times a day) to the official Clojure tap https://github.com/clojure/homebrew-tools as @ formulas. Much less frequently, we mark a release as "stable" and the stable version can always be found at https://github.com/clojure/homebrew-tools/blob/master/Formula/clojure.rb (installed with `brew install clojure/tools/clojure`).

The clojure formula here has been updated several times recently to newer prerelease development versions which are buggy or have as yet unannounced changes. We are now fielding regular support questions related to these prerelease versions because these versions of the tool are not ready for widespread use yet. This PR sets the clojure version backwards to the current stable version.

Is there a better way to stop this from happening in the future? Is there some way to notify users in the formula description etc that this is not the official Clojure formula supported by the Clojure team? Would it make sense to remove the formula and have a message that pointed people to the official Clojure tap instead?

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by chenrui333** [comment]
So technically, this would be a downgrade, is there anyway that we can track down the release other than using brew tap version declaration?

```
==> brew audit clojure --online --git --skip-style
==> FAILED
Error: 1 problem in 1 formula detected
clojure:
  * stable version should not decrease (from 1.10.1.645 to 1.10.1.561)
```

**Comment by chenrui333** [comment]
> The clojure formula here has been updated several times recently to newer prerelease development versions which are buggy or have as yet unannounced changes. We are now fielding regular support questions related to these prerelease versions because these versions of the tool are not ready for widespread use yet. This PR sets the clojure version backwards to the current stable version.

We have `livecheck` dev-cmd that maintainer used for keeping the software up to date.

> Is there a better way to stop this from happening in the future? Is there some way to notify users in the formula description etc that this is not the official Clojure formula supported by the Clojure team? Would it make sense to remove the formula and have a message that pointed people to the official Clojure tap instead?

We can technically remove this from the core in favor of the formula of the tap one.

**Comment by chenrui333** [comment]
The issue that I can see is there is no good way to tell which version is the `latest` stable releasse.

**Comment by chenrui333** [comment]
Another advantage that we have the formula in the core is to support clojure tools ecosystem (from what I can see, `Planck` replies on `clojure` for build)

**Comment by puredanger** [comment]
Updating livecheck to only look at the stable formula seems like a good solution, thanks

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
@puredanger, honestly, I can still see the value of keeping the tap in the core (essentially considering clojure is a major programming language), it would be cross-tested with the clojure ecosystem at large. 

I did read your post about the hassles with the core, but based on my experience, I dont see any issues at all, we are reacting on the PRs and issues fairly quickly. :)

**Comment by puredanger** [comment]
The biggest thing is that here anyone change stuff at any time (and they have, and in ways we disagree with), and that's fundamentally at odds with our desire to control what we release. I do appreciate your help on this PR, and I think this is a good path forward. Thanks!

**Comment by chenrui333** [comment]
> The biggest thing is that here anyone change stuff at any time (and they have, and in ways we disagree with), and that's fundamentally at odds with our desire to control what we release. I do appreciate your help on this PR, and I think this is a good path forward. Thanks!

That is actually a good comment, maybe we should have some mechanism to share the responsibility for the individual formulae to make folks comfortable to rollout the upgrades. Currently, it is the maintainer the biggest drive to keep the formulae up to date (which might incurs some sort of controversies down the path), I sincerely think the shared model would be more sustainable and healthy.

**Comment by MikeMcQuaid** [comment]
> Much less frequently, we mark a release as "stable"

How do you do this? Is this a GitHub "prerelease" label? Is it something in the version number? How can we programatically query this information?

> The biggest thing is that here anyone change stuff at any time (and they have, and in ways we disagree with), and that's fundamentally at odds with our desire to control what we release.

You are in complete control in what you release. If you make releases that you don't want people to use: perhaps you should not make those releases?

Homebrew has been, is and (as long as I have any influence) always will be run by and for the community. If you wish things to not be changed you need to work with us to find programatic solutions e.g. `test do` block additions, `brew audit`s etc. 

Note: this is how all system open source distributions work. If you release your code in the open under an open source license: you are saying we can package and ship whatever version we see fit.

> maybe we should have some mechanism to share the responsibility for the individual formulae to make folks comfortable to rollout the upgrades.

The maintainers of Homebrew are responsible for maintaining the core taps of Homebrew. We do not (and should not) defer to upstreams desire to have "more control" to the detriment of our users and workflows.

**Pull request merged**
